### PR TITLE
import-json

### PR DESCRIPTION
**Pull request by mbredif**
This PR adds a new entrypoint to import json files that already follow the li3ds data model. The json structure must be a dictionary that maps object types to an array of JSON objects. The only specificity is that ids (both definitions and references) are for now local to the json file, so that li3ds objects with a imported json file may reference each other. These local ids are not preserved once the objects are imported in the database and neither can they reference ids of object that are already in the database.

```
{
"sensor": [ ... sensors... ],
"referential": [ ... referentials ...],
"transfos/type": [ ... transfo types ... ],
"transfo": [ ... transfos ... ],
"transfotree": [ ... transfo trees ... ],
"platform": [ ... platforms ... ],
"platforms/{id}/config": [ ... configs ... ],
"project": [ ... projects ... ],
"session": [ ... sessions ... ],
"datasource": [ ... datasources ... ]
}
```

The rationale of this PR is to enable to insert data within the li3ds database directly (ie without converting from any other datamodel).

### CODE REVIEW

File: cli_li3ds/import_json.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import json
+import pathlib
+
+from cliff.command import Command
+
+from . import api
+
+
+class ImportJson(Command):
+    """ import one or several images
```
Comment (elemoine) [inline cli_li3ds/import_json.py]: Please change this.

File: cli_li3ds/import_json.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    """
+
+    log = logging.getLogger(__name__)
+
+    image_date_cache = {}
```
Comment (elemoine) [inline cli_li3ds/import_json.py]: To be removed.

File: cli_li3ds/import_json.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    def get_parser(self, prog_name):
+        self.log.debug(prog_name)
+        parser = super().get_parser(prog_name)
+        api.add_arguments(parser)
+        parser.add_argument(
+            '--json-dir', '-f', default='.',
+            help='base directory to search for json files (optional, default is ".")')
+        parser.add_argument(
+            '--uri', default='{uri}',
+            help='uri pattern for datasources (optional, default is path through "{uri}")')
```
Comment (elemoine) [inline cli_li3ds/import_json.py]: Do you mean "pass through" here?

File: cli_li3ds/import_json.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                projects = self.handle_json(objs, json_path, uri, api.Project)
+                sessions = self.handle_json(objs, json_path, uri, api.Session, projects, platforms)
+                self.handle_json(objs, json_path, uri, api.Datasource, sessions, referentials)
+
+        objs.get_or_create()
+        self.log.info('Success!\n')
+
+    @classmethod
+    def handle_json(cls, objs, json_path, uri, class_, deps1=None, deps2=None):
+        with json_path.open(encoding="iso-8859-1") as f:
```
Comment (elemoine) [inline cli_li3ds/import_json.py]: iso-8859-1? The json files are guaranteed to be iso-8859-1 encoded?

File: cli_li3ds/api.py [hunk 5]
```diff
@@ -616,4 +618,6 @@ def update_obj(args, metadata, obj, type_):
 
 
 def isoformat(date):
+    if isinstance(date, str):
+        date = dateutil.parser.parse(date)
```
Comment (elemoine) [inline cli_li3ds/api.py]: Python has functions for parsing dates. See https://github.com/LI3DS/cli-li3ds/blob/fc65f28ad09cb1e340e61328bd59b13876e101f8/cli_li3ds/import_image.py#L187-L189 for example. Are you sure we need an extra dependency for that?

**Review by elemoine — CHANGES_REQUESTED** [review]
Looks good. Just a few requested changes.

**Comment by elemoine** [comment]
Also, it would be good to add a test file and have test.sh import it.

File: cli_li3ds/import_json.py (hunk 1 above)
Comment (mbredif) [reply cli_li3ds/import_json.py]: ok

File: cli_li3ds/import_json.py (hunk 2 above)
Comment (mbredif) [reply cli_li3ds/import_json.py]: ok

File: cli_li3ds/import_json.py (hunk 3 above)
Comment (mbredif) [reply cli_li3ds/import_json.py]: yes

File: cli_li3ds/import_json.py (hunk 4 above)
Comment (mbredif) [reply cli_li3ds/import_json.py]: I just know importing https://github.com/LI3DS/cli-li3ds/blob/import-json/data/2017_17FA7506_C_6.json did not work without specifying this encoding. 
Do you have a better proposition ?

File: cli_li3ds/api.py (hunk 5 above)
Comment (mbredif) [reply cli_li3ds/api.py]: Accuracy in minutes is not sufficient for json import of `_time` parameters.
I was looking to parse iso formatting dates and got [here](https://stackoverflow.com/questions/127803/how-to-parse-an-iso-8601-formatted-date), which hints at [dateutil](https://dateutil.readthedocs.io/en/stable/parser.html) or [iso8601](http://pyiso8601.readthedocs.io/en/latest/)

**Comment by mbredif** [comment]
I added import-json to test.sh and took into account minor comments. I am not sure what is the right thing to do for the encoding and isoformat remarks.

**Review by elemoine — APPROVED** [review]

**Pull request merged**